### PR TITLE
Move GetEntityTypeIDsFromProperties to EntityTypeManager

### PR DESCRIPTION
**Pull request by scober**
This function is useful outside of PropertyGraph and it initializes an
EntityTypeManager. So a sensible home for it is as a static function in
the EntityTypeManager class.

One downside of this move is that we have to de-parallelize the function.
EntityTypeManager is in libsupport and libsupport doesn't have access to
the libgalois thread pool.

### CODE REVIEW

File: libsupport/include/katana/EntityTypeManager.h
Comment (witchel) [inline libsupport/include/katana/EntityTypeManager.h]: This is a trivial comment, but functions that start with `Get `tend to be simple accessors.  This looks more like `RebuildEntityTypeIDsFromProperties`.

Comment (amberhassaan) [reply libsupport/include/katana/EntityTypeManager.h]: I agree. Good suggestion.

Comment (scober) [reply libsupport/include/katana/EntityTypeManager.h]: My current draft is `AssignEntityTypeIDsFromProperties()`

File: libsupport/include/katana/EntityTypeManager.h
Comment (amberhassaan) [inline libsupport/include/katana/EntityTypeManager.h:87]: We should `PropertyColumn` private. May be even local to the function.

File: libsupport/include/katana/EntityTypeManager.h
Comment (amberhassaan) [inline libsupport/include/katana/EntityTypeManager.h:104]: template template types usually cause more problems than they solve. My recommendation is to just use `typename EntityTypeVec` and then you can `static_assert(std::is_same_v<typename EntityTypeVec::value_type, EntityTypeID>);`

File: libsupport/include/katana/EntityTypeManager.h
Comment (amberhassaan) [inline libsupport/include/katana/EntityTypeManager.h:105]: Might wanna move this big ass function into a cpp. In fact, in a later PR, I might break it up further just because I don't like long methods.

File: libsupport/include/katana/EntityTypeManager.h
Comment (amberhassaan) [inline libsupport/include/katana/EntityTypeManager.h:262]: Please see if you can remove this method while you're at it.

File: libsupport/include/katana/EntityTypeManager.h
Comment (roshandathathri) [inline libsupport/include/katana/EntityTypeManager.h:104]: Is a template parameter required? If so, consider instantiating the specific ones required. This function is big enough that it's better defined in a cpp file. You can do this in a subsequent PR.

File: libsupport/include/katana/EntityTypeManager.h
Comment (arthurp) [inline libsupport/include/katana/EntityTypeManager.h:121]: This needs to return an errors instead of being an assert since a caller can trigger this failure.

**Review by roshandathathri — APPROVED** [review]
LGTM

File: libsupport/include/katana/EntityTypeManager.h
Comment (arthurp) [inline libsupport/include/katana/EntityTypeManager.h:105]: Why are you moving this to libsupport? It seems odd since libsupport is not really "property-aware" and just deals with low-level vectors and arrays for the most part.

Where else do you expect this to be used other than with PropertyGraph?

File: libtsuba/src/RDGSlice.cpp
Comment (arthurp) [inline libtsuba/src/RDGSlice.cpp:44]: Given that there are some comments that may need addressing and @amberhassaan seemed to be commenting as well, please split out this change into a separate PR to get it in quicker.

This kind of fix should generally be alone in a PR anyway, so as to unblock CI as soon as possible.

**Review by arthurp — APPROVED** [review]
Please look at my comments. I am a bit concerned about moving that function.

That said, I would be OK...


Well you merged it. :-) That works too. ;-)

These are comments for the future then.

File: libsupport/include/katana/EntityTypeManager.h
Comment (scober) [reply libsupport/include/katana/EntityTypeManager.h:104]: The template was a workaround for handling `NUMAArray`. I don't think we need to support arrays other than `NUMAArray`, but as long as we define this in `libsupport` I don't think we can instantiate that instance.

File: libsupport/include/katana/EntityTypeManager.h
Comment (scober) [reply libsupport/include/katana/EntityTypeManager.h:121]: Will do

File: libtsuba/src/RDGSlice.cpp
Comment (amberhassaan) [inline libtsuba/src/RDGSlice.cpp:33]: Having to do this math is laborious and error prone. Consider adding a TODO where this interface can be improved to handle data type sizes automatically.

**Review by amberhassaan — APPROVED** [review]
Minor comments.

File: libsupport/include/katana/EntityTypeManager.h
Comment (roshandathathri) [reply libsupport/include/katana/EntityTypeManager.h:104]: I understand the point. I'm a bit concerned about increased compile time because this header will be included in many places. Can we instead take a C-style pointer to represented an array? I'm not sure if that is better for code quality but we don't expect this function to be used a lot anyway.

Comment (amberhassaan) [reply libsupport/include/katana/EntityTypeManager.h:104]: You know, we could solve this issue by introducing a `Buffer` type that takes a pointer and acts like a vector/array and has a size. Pretty much an Arrow Buffer (but without the unnecessary shared_ptr).

Comment (arthurp) [reply libsupport/include/katana/EntityTypeManager.h:104]: This function is currently only called from PropertyGraph.cpp. I honestly think it should probably be there. It's pretty coupled to PropertyGraph loading.

File: libsupport/include/katana/EntityTypeManager.h
Comment (scober) [reply libsupport/include/katana/EntityTypeManager.h:105]: I am not exactly sure what you mean here, but I will try to address your concerns. libsupport is now type-aware and due to our previous type representation, being type-aware requires a certain amount of property-awareness.

I don't actually need this function to be in libsupport, I just need it to be publicly accessible. Because one purpose of this function is to initialize an `EntityTypeManager`, I thought this would be a natural home for it. This seemed like it would be an awkward appendage to `PropertyGraph`.

The high level answer to your second question is that this function is required anywhere we need to update an old-style graph to a new-style graph. The specific answer is that I am working on a code path in the enterprise repo that loads graphs without creating `PropertyGraph`s.

File: libsupport/include/katana/EntityTypeManager.h
Comment (scober) [reply libsupport/include/katana/EntityTypeManager.h:87]: Will do

File: libsupport/include/katana/EntityTypeManager.h
Comment (roshandathathri) [reply libsupport/include/katana/EntityTypeManager.h:104]: Amber, that's a good idea.

Arthur, no, it needs to be called when the RDG is loaded. That PR will come next.

Comment (scober) [reply libsupport/include/katana/EntityTypeManager.h:104]: Interesting. I was not aware of that pattern. I will give it a shot in the follow-up PR.

File: libsupport/include/katana/EntityTypeManager.h
Comment (scober) [reply libsupport/include/katana/EntityTypeManager.h:105]: The issue with putting it in a cpp is that it is templated and the type that we really care about passing as the template parameter is in libgalois. I will see what I can do to break out some helpers and put them in a cpp.

File: libsupport/include/katana/EntityTypeManager.h
Comment (scober) [reply libsupport/include/katana/EntityTypeManager.h:262]: Remove the `Reset()` method?

File: libtsuba/src/RDGSlice.cpp
Comment (scober) [reply libtsuba/src/RDGSlice.cpp:33]: Improve the interface to `FileView`?

File: libsupport/include/katana/EntityTypeManager.h
Comment (scober) [reply libsupport/include/katana/EntityTypeManager.h:104]: I will add a `Buffer` class to libsupport in my follow-up PR.

Comment (arthurp) [reply libsupport/include/katana/EntityTypeManager.h:104]: I would like to see benchmarks that show the downside of using arrow::Array before we create our own copy. I suspect the overhead of `shared_ptr` is lower than people think in any real context. I may well be wrong. But adding our own copies of things before benchmarking seems silly and a good way to force ourselves to reimplement more and more of arrow.

Comment (scober) [reply libsupport/include/katana/EntityTypeManager.h:104]: @arthurp Are you suggesting that we take an `arrow::Array` as the parameter here?

Comment (arthurp) [reply libsupport/include/katana/EntityTypeManager.h:104]: I was thinking `arrow::Buffer` but `arrow::Array` would be fine too in terms of what I'm thinking about.

Comment (witchel) [reply libsupport/include/katana/EntityTypeManager.h:104]: Cool trick.

Comment (scober) [reply libsupport/include/katana/EntityTypeManager.h:104]: Is `arrow::Buffer` designed to be mutable?

Comment (arthurp) [reply libsupport/include/katana/EntityTypeManager.h:104]: Yes.

File: libsupport/include/katana/EntityTypeManager.h
Comment (amberhassaan) [reply libsupport/include/katana/EntityTypeManager.h:105]: yes, I did realize it later on that it's a template function.

File: libsupport/include/katana/EntityTypeManager.h
Comment (amberhassaan) [reply libsupport/include/katana/EntityTypeManager.h:262]: yes. I don't think we need it. Any use can replaced with `= EntityTypeManager{}`

File: libtsuba/src/RDGSlice.cpp
Comment (amberhassaan) [reply libtsuba/src/RDGSlice.cpp:33]: yes. My point being that having to do this math is error prone. It's like calling `malloc` vs `new` and people forget to multiply length with `sizeof(T)`.

File: libsupport/include/katana/EntityTypeManager.h
Comment (amberhassaan) [reply libsupport/include/katana/EntityTypeManager.h:104]: Does `arrow::Buffer` have the coveted `operator []`? Secondly, if I don't need a `shared_ptr`, then I shouldn't use it. I just need to wrap a pointer into a `Buffer`. `Buffer` never owns the memory nor frees it. So I'm in favor of adding our own simple `Buffer` types and I'll even volunteer to add one. I've had this itch for a while.

Comment (amberhassaan) [reply libsupport/include/katana/EntityTypeManager.h:104]: I looked it up and `arrow::Buffers` are untyped, so totally unfit for our use case here. I want something typed that behaves like an array or vector.

File: libtsuba/src/RDGSlice.cpp
Comment (scober) [reply libtsuba/src/RDGSlice.cpp:33]: I did some work on `FileView` about a year ago and my thoughts then and now are that we should offer a few flavors of `FileView`. It currently serves several purposes and it doesn't make sense to add special functionality for each of them. For example, it doesn't make sense to have a `Bind()` variant that converts number_of_elements -> number_of_bytes for an object that is used to read Parquet files. Just like it doesn't make sense to have prefetching for an object that is used to read slices of binary arrays.

But if we had a `ParquetView`, a `BinaryArrayView`, and a `CSRView` that each maintained a `FileView` internally (via subclassing or pointers), we could have good interfaces for all of those use cases and put only truly common functionality in `FileView`.

Comment (amberhassaan) [reply libtsuba/src/RDGSlice.cpp:33]: I see. I'll leave it up to you to decide how to proceed. I won't be touching `FileView` in the near future (not yet).

**Pull request merged**
